File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int maxwater=0,n=height.length;
        int left=0,right=n-1,leftmax=height[0],rightmax=height[n-1];
        while(left<right){
            leftmax=Math.max(leftmax,height[left]);
            rightmax=Math.max(rightmax,height[right]);
            if(leftmax<rightmax){
                maxwater+=leftmax-height[left++];
            }
            else{
                maxwater+=rightmax-height[right--];
            }
        }
        return maxwater;
    }
} 

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i=0
        leftmax=height[0]
        sum=0
        j=len(height)-1
        rightmax=height[j]
        while i<j:
            if leftmax <= rightmax:
                sum+=leftmax-height[i]
                i+=1
                leftmax=max(leftmax,height[i])
            else:
                sum+=rightmax-height[j]
                j-=1
                rightmax=max(rightmax,height[j])
        return sum